src/jitlog/generator.lua

```diff
@@ -339,7 +333,431 @@ function parser:complete()
   return data
 end
 
+local generator = {}
+
+local function joinlist_format(list, fmt, prefix, suffix, addmax)
+  prefix = prefix or ""
+  suffix = suffix or ""
+  local t = {}
+  for i, name in ipairs(list) do
+    t[i] = format(fmt, name)
+  end
+  if addmax then
+    table.insert(t, format(fmt, "MAX"))
+  end
+  return prefix .. table.concat(t, suffix .. prefix) .. suffix
+end
+
+local unescapes = {
+  n = '\n',
+  s = ' ',
+  t = '\t',
+}
+
+local function unescape(s)
+  return string.gsub(s, "\\(%a)", function(key)
+    local result = unescapes[key]
+    assert(result, "unknown escape")
+    return result
+  end)
+end
+
+function generator:buildtemplate(tmpl, values)
+  return string.gsub(tmpl, "{{(.-)}}", function(key)
+    local name, fmt = string.match(key, "%s*(.-)%s*:(.+)")
+    if name then
+      key = name
+      fmt = unescape(fmt)
+    end
+    
+    local value = values[key]
+
+    if value == nil then
+      error("missing value for template key '"..key.."'")
+    end
+    
+    if fmt then
+      if type(value) == "table" then
+        value = joinlist_format(value, fmt)
+      else
+        value = format(fmt, value)
+      end
+    end
+    return value
+  end)
+end
+
+function generator:write(s)
+  --print(s)
+  self.outputfile:write(s)
+end
+
+function generator:writef(s, ...)
+  assert(type(s) == "string" and s ~= "")
+  self.outputfile:write(format(s, ...))
+end
+
+function generator:writetemplate(name, ...)  
+  local template = self.templates[name]
+  local result
+  if not template then
+    error("Missing template "..name)
+  end
+  if type(template) == "string" then
+    assert(s ~= "")
+    result = self:buildtemplate(template, ...)
+  else
+    result = template(...)
+  end
+  
+  self:write(result)
+end
+
+function generator:mkfield(f)
+  local ret, getter
+  
+  local comment_line = self.templates.struct_comment or self.templates.comment_line
+
+  if f.type == "bitfield" then
+    ret = format("\n/*  %s: %d;*/", f.name, f.bitsize)
+  else
+    local type = self.types[f.type]
+    local langtype = type[self.outputlang] or type.c
+
+    if f.type == "string" and self.templates.stringfield then
+      ret = format(self.templates.stringfield, f.name)
+    elseif  f.vlen then
+      ret = "\n"..format(comment_line, format("%s %s[%s];", langtype, f.name, f.buflen))
+    elseif f.bitstorage or f.type == "string" then
+      -- A bit length integer stuff in the spare bits of the id field
```

Comment:
I can't parse the comment, and I'm not quite sure how it relates to having a string type?

Reply:
It writing comments in place of field entry in a message struct declaration that can't be directly declared in target language like C things like variable length arrays and strings. It was from a time when handling for strings had to be specialcased more.  I think the previous elseif check is now catching strings as well so it should be removed

Reply:
fixed and cleaned up another dead if in the function